### PR TITLE
Speed up DeepSeek V4.1 Flash on Metal

### PR DESCRIPTION
**Pull request by kernelpool**
# Summary

This PR adds optimizations and DSpark support to the DeepSeek V4.1 Flash engine for Metal. All benchmarks and evals were measured on two Mac Studio M3 Ultra (512 GB). DSpark seems to work well against code (for the context lengths tested, up to 256k), but prose still needs some more tuning.

Note that there are some pending work/optimizations still being done so I invite people to contribute.

# Test setup

- Machines: 2x Mac Studio M3 Ultra, 32-core CPU, 80-core GPU, 512 GB
- Base: `main` 8db1d1d, built from the same tree state for every "main" number below.
- Models: [DeepSeek-V4.1-Flash-MXFP4](https://huggingface.co/kernelpool/DeepSeek-V4.1-Flash-MXFP4-GGUF), DeepSeek-V4.1-Flash-Q4
- Conversion:

```sh
make -C gguf-tools libds4quants.dylib
python3 gguf-tools/deepseek41_quantize.py \
  --hf models/DeepSeek-V4.1-Flash \
  --source-revision dba1be0a40aa45a94ad051997016db3960a90277 \
  --quant mxfp4 \
  --out gguf/DeepSeek-V4.1-Flash-MXFP4.gguf \
  --dspark-out gguf/DeepSeek-V4.1-Flash-DSpark-MXFP4.gguf
```

# Speed

## Standard sweep

```sh
./ds4-bench -m MODEL.gguf --metal --prompt-file speed-bench/promessi_sposi.txt \
  --ctx-start 2048 --ctx-max 65536 --step-incr 2048 --gen-tokens 128 --csv OUT.csv
```

Generation t/s (incremental 2048-token prefill t/s in parentheses):

| Context | main, MXFP4 | this branch, MXFP4 | main, Q4 | this branch, Q4 | main, Q2 | this branch, Q2 |
|---|---:|---:|---:|---:|---:|---:|
| 2048 | 19.2 (329) | 34.7 (368) | 16.9 (330) | 33.2 (396) | 19.7 (340) | 36.1 (385) |
| 16384 | 19.0 (317) | 33.7 (335) | 16.8 (317) | 32.3 (363) | 19.8 (332) | 35.1 (358) |
| 32768 | 18.6 (303) | 32.9 (314) | 16.4 (301) | 31.7 (341) | 19.4 (313) | 34.9 (337) |
| 65536 | 18.2 (290) | 31.3 (276) | 16.2 (269) | 30.4 (302) | 18.8 (279) | 34.2 (298) |

## DSpark on English and Code (XNU)

```sh
./ds4-bench -m DeepSeek-V4.1-Flash-MXFP4.gguf --metal --prompt-file PROMPT.txt \
  --ctx-start 4096 --ctx-max 262144 --step-mul 2 --gen-tokens 256 --csv OUT.csv \
  [--dspark --mtp-model DeepSeek-V4.1-Flash-DSpark-MXFP4.gguf]
```

| Context | XNU plain | XNU DSpark | prose plain | prose DSpark |
|---|---:|---:|---:|---:|
| 4096 | 34.7 | 59.2 | 34.7 | 35.0 |
| 8192 | 34.5 | 37.5 | 34.5 | 33.9 |
| 16384 | 33.9 | 53.1 | 34.0 | 33.0 |
| 32768 | 33.2 | 46.4 | 33.2 | 32.5 |
| 65536 | 31.8 | 38.8 | 31.8 | 30.9 |
| 131072 | 29.3 | 45.8 | 29.2 | 28.7 |
| 262144 | 25.3 | 50.4 | - | - |

XNU: 1.63 MB of osfmk/kern C sources. 
Prose: an English novel (Pride and Prejudice)

## Server sweep

Tests against `ds4-server`, 256 output tokens, 2 runs, default sampling (prompt t/s over generation t/s):

| Context | main, Q4 | this branch, Q4 | this branch, MXFP4 | main, Q4, TP | this branch, MXFP4, TP |
|---|---:|---:|---:|---:|---:|
| 4k | 326 / 16.9 | 474 / 32.7 | 477 / 34.3 | 320 / 22.0 | 496 / 41.4 |
| 8k | 328 / 16.7 | 586 / 32.5 | 591 / 33.9 | 322 / 22.0 | 604 / 41.2 |
| 16k | 586 / 16.7 | 658 / 32.3 | 669 / 33.6 | 559 / 21.8 | 673 / 40.7 |
| 32k | 676 / 16.6 | 719 / 32.0 | 752 / 33.3 | 553 / 21.5 | 694 / 40.2 |
| 64k | 711 / 16.4 | 680 / 30.4 | 683 / 31.5 | 613 / 21.0 | 705 / 39.3 |
| 128k | 671 / 15.6 | 575 / 28.3 | 602 / 30.0 | 617 / 20.0 | 709 / 38.1 |
| 256k | 588 / 14.8 | 555 / 27.5 | 592 / 29.4 | 561 / 18.3 | 643 / 36.3 |

# Quality

## Evals

```sh
./ds4-eval -m MODEL.gguf --metal --plain -c 32768 --temp 0 --questions 20
```

Core suite, first 20 questions (wall time in parentheses):

| Model | main | this branch |
|---|---|---|
| MXFP4 | 19/20 (43 min) | 19/20 (21 min) |
| Q4 | 20/20 (39 min) | 20/20 (20 min) |
| Q2 | 19/20 (62 min) | 19/20 (30 min) |

`gguf-tools/quality-testing/score_official` on the tracked V4.1 Flash sets (avg_nll, API top-1 agreement):

| Set | Q4 | MXFP4 | Q2 |
|---|---:|---:|---:|
| general, 100 | 0.2524, 96.7% | 0.2410, 97.0% | 0.3657, 90.3% |
| sparse boundary, 17 | 0.1887, 97.9% | 0.1965, 97.9% | 0.4159, 90.8% |
| batched prefill, 12 | 0.3401, 95.3% | 0.3379, 97.0% | 0.5312, 86.6% |
| long, 9 | 0.3288, 93.4% | 0.3356, 94.8% | 0.5571, 85.2% |
| extended, 6 | 0.3665, 93.8% | 0.3555, 95.6% | 0.5586, 85.2% |

### CODE REVIEW

**Comment by jjakemaness** [comment]
Results from a smaller configuration than the ones in the description, in case they're useful: **one Mac Studio M3 Ultra, 80-core GPU, 256 GB, macOS 27.0**, `DeepSeek-V4.1-Flash-Q2.gguf` (source revision `df42c109`), resident, `--ctx 262144`, temperature 0.

**Plain decode — a clear win.** Same machine, same OS, same Q2 file, same prompt, 200 tokens:

| build | tok/s |
| --- | --- |
| ivanfioravanti `ds41f-dspark` + ungated parallel Engram reads | 30.52, 30.59, 30.53 |
| this PR | **37.12, 37.10, 37.08** |

Generated text was byte-identical between the two. Single runs at larger inputs: ~10k-token prompt 36.1 tok/s; a one-shot ~115k-token prompt, prefill 580 t/s and generation 32.0 tok/s.

In real agent use (`ds4-agent`, reading five C files totalling ~7.6k lines / ~80k tokens in chunked tool reads, then answering) it completed cleanly in 465 s.

**DSpark — fast, but it stalled for me on macOS 27.** With a support file converted by this PR's tool (`--quant mxfp4 --dspark-out`):

- short prompt: 40.8, 40.7 tok/s warm (plain 37.2); ~10k prompt: 47.2 tok/s (plain 36.1) — single run
- the same five-file agent task **hung** about three minutes in, while ingesting a ~600-line read with roughly 40–50k tokens of history; no progress for 3+ minutes before I stopped it. The identical task without DSpark finished.
- a one-shot ~115k-token prompt with DSpark never reached its first 32k progress mark in 8 minutes (twice, with and without `DS4_METAL_NO_RESIDENCY`). Without DSpark it completed earlier the same day.

These are single observations, not a diagnosis. Happy to run anything that would help narrow it down.

**Tip: building the DSpark support file without all 48 shards.** The converter builds the full-model plan even for a `--dspark-out`-only run, so it wants every shard present — ~330 GB. The DSpark tensors only come from shards 44–46. I satisfied the check with header-only placeholders for the other 45: range-request each shard's safetensors header (the first 8 bytes give its length), write the header, then `truncate` the file to its real size so it's sparse. About 10 MB of real disk. Before converting I checked the dry-run plan to confirm all 81 draft tensors source from shards 44–46, so nothing reads the placeholder zeros.

**Comment by jjakemaness** [comment]
Correction to my comment above: **the DSpark stalls were caused by my own launch settings, not by this PR.**

I was running with `DS4_METAL_NO_RESIDENCY=1` and `DS4_METAL_DISABLE_QUEUE_KEEPALIVE=1` (set to release RAM when idle). With both removed, i.e. the defaults, the same tests pass with DSpark on (this PR at `fd70919`, same Mac, same Q2 file):

| test | tok/s |
|---|---|
| short prompt | 40.8, 40.9 (plain: 37.7–37.9) |
| ~10k prompt | 47.4, 47.2 |
| one-shot ~115k prompt | 34.4, completed |
| five-file agent task + 1 min idle + 8,437-line file read, to 219.7k ctx | completed, twice |

Through `ds4-server` with the disk KV cache, a follow-up in the same 115k conversation reused 114,810 of 114,826 tokens and answered in 2 s.

When it hung, a sample showed the worker in `waitUntilCompleted` from `ds41_graph_prefill_sweep`, and the Metal dispatch thread blocked inside `-[IOGPUMetalCommandQueue submitCommandBuffers:count:]` with the GPU at 0%. That's consistent with the driver rebuilding its mappings of the ~155 GB no-copy model views after the queue sat idle, the case the queue-keepalive comment in `ds4_metal.m` describes. On macOS 27 that rebuild appears to take minutes rather than under a second.

I also compared the latest two commits (`af7c02b`) against `fd70919` on this setup: no measurable difference at short or 10k, and within run-to-run noise at 115k.

Sorry for the noise. The plain-decode numbers in my first comment stand.

**Comment by kernelpool** [comment]
Thanks for the feedback! Made the proposed change to the converter.

**Comment by jjakemaness** [comment]
No prob! Got this running super fast on 256gb m3 ultra now! 115k prompt mid 30s, 10k prompt 40s, short prompt 48-50 tok/s!

**Comment by triplesparkle** [comment]
@kernelpool - I built your branch at `a15028ee4` on a DGX Spark GB10 against `main` at
`8db1d1d1`. Nice work. Three findings, one of which you will want.

**Builds clean on CUDA now**, `rc=0` in 205 s: `4fe6a7802` and `ed43e1a56` closed the two failures
I patched around a few days ago.

**Decode ties here**, `+0.21 %` against this round's own `2.27 %` floor, both arms at the same
90GB cache and 8,941 delivered slots. No surprise, the fused work is Metal and decode on this box
is bandwidth-bound.

**Greedy output differs from main, gated by prompt length.** Same text, growing prefix, 4 greedy
tokens:

| prompt tokens | 81 | 280 | 1,040 | 4,391 | 17,706 |
|---|---:|---:|---:|---:|---:|
| median top-20 logit delta | 1.560051 | 0.000000 | 0.000000 | 0.058922 | 0.000000 |

Three exact zeros, and main against main is `0.000000`, so not float noise. A short prompt flips a
token outright: index 6 of 32, main takes `.` over ` why` by 0.904201 logits, yours reverses it. A
long one only nudges the distribution, which is probably why it has gone unseen. 20 seconds:

```
head -c 200 speed-bench/promessi_sposi.txt > /tmp/p200.txt
./ds4 --cuda -m <V4.1 Flash Q2> --prompt-file /tmp/p200.txt --temp 0 -n 4 --dump-logprobs
```

Small one: `4fe6a7802` left `ds41_router_one_off`, `ds41_concurrent_ffn_on` and
`ds41_tp_balanced_views` with every caller inside `#ifdef __APPLE__`, so a CUDA build warns on all
three plus an unused `gate` in `ds41_sum_partial_batch`. Main warns zero.

All of it is on `TripleSparkleAI/ds4:triple-kernelpool-spark`. Happy to run anything else on the
Spark.

**Comment by ivanfioravanti** [comment]
Amazing job on Apple Silicon side! I'm gonna add a very small PR to boost a little smaller contexts up to 16K, for the rest it was just perfect! @kernelpool you rock!
